lib/instrumentation/modules/aws-sdk/s3.js

```diff
@@ -0,0 +1,127 @@
+//
+// For example: 'headBucket' -> 'HeadBucket'
+function opNameFromOperation (operation) {
+  return operation[0].toUpperCase() + operation.slice(1)
+}
+
+// Return an APM "resource" string for the bucket, Access Point ARN, or Outpost
+// ARN. ARNs are normalized to a shorter resource name.
+//
+// Known ARN patterns:
+// - arn:aws:s3:<region>:<account-id>:accesspoint/<accesspoint-name>
+// - arn:aws:s3-outposts:<region>:<account>:outpost/<outpost-id>/bucket/<bucket-name>
+// - arn:aws:s3-outposts:<region>:<account>:outpost/<outpost-id>/accesspoint/<accesspoint-name>
+//
+// In general that is:
+//    arn:$partition:$service:$region:$accountId:$resource
+//
+// This parses using the same "split on colon" used by the JavaScript AWS SDK v3.
+// https://github.com/aws/aws-sdk-js-v3/blob/v3.18.0/packages/util-arn-parser/src/index.ts#L14-L37
+function resourceFromBucket (bucket) {
+  let resource = null
+  if (bucket) {
+    resource = bucket
+    if (resource.startsWith('arn:')) {
+      resource = bucket.split(':').slice(5).join(':')
+    }
+  }
+  return resource
+}
+
+// Instrument an awk-sdk@2.x operation (i.e. a AWS.Request.send or
+// AWS.Request.promise).
+//
+// @param {AWS.Request} request https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Request.html
+function instrumentationS3 (orig, origArguments, request, AWS, agent, { version, enabled }) {
+  const opName = opNameFromOperation(request.operation)
+  let name = 'S3 ' + opName
+  const resource = resourceFromBucket(request.params && request.params.Bucket)
+  if (resource) {
+    name += ' ' + resource
+  }
+
+  const span = agent.startSpan(name, TYPE, SUBTYPE, opName)
+  if (span) {
+    request.on('complete', function onComplete (response) {
+      // `response` is an AWS.Response
+      // https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Response.html
+
+      // Determining the bucket's region.
+      // `request.httpRequest.region` isn't documented, but the aws-sdk@2
+      // lib/services/s3.js will set it to the bucket's determined region.
+      // This can be asynchronously determined -- e.g. if it differs from the
+      // configured service endpoint region -- so this won't be set until
+      // 'complete'.
+      const region = request.httpRequest.region
+
+      // Destination context.
+      // '.httpRequest.endpoint' might differ from '.service.endpoint' if
+      // the bucket is in a different region.
+      const endpoint = request.httpRequest.endpoint
+      const destContext = {
+        address: endpoint.hostname,
+        port: endpoint.port,
+        service: {
+          name: SUBTYPE,
+          type: TYPE
+        }
+      }
+      if (resource) {
+        destContext.service.resource = resource
+      }
+      if (region) {
+        destContext.cloud = { region }
+      }
+      span.setDestinationContext(destContext)
+
+      if (response) {
```

Comment:
I opened https://github.com/elastic/apm-agent-nodejs/issues/2125 to drop the HTTP spans under S3 spans (similar to the issues we have for ES and SQS already). Per my understanding of the current specs and recent discussion we want the AWS spans to be exit spans, not have the underlying HTTP spans, and the spec requires that exit spans set outcome.

I have some open Qs that I'm chatting with @estolfo about currently that will lead to a PR against the specs suggesting that `context.http.status_code` be added to AWS spans, if feasible.

The best argument for setting outcome based on status code is an S3 GetObject call that uses the available options to do a conditional request, e.g. `IfNoneMatch: $etag`, for which you get an *error* in the callback from the S3 client, but a 304 Not Modified response. That should be a "success". I discussed that with @estolfo  as well and likely the coming spec PR will clarify that. Her understanding was that the Ruby agent already did this automatically because if their HTTP-layer instrumentation.  FWIW, I think the other Node.js agent `aws-sdk` instrumentations should eventually share this handling of the `function (err, data)` callback to set outcome based on statusCode if available.